main.py

```diff
@@ -718,43 +736,46 @@ def start_active_view():
             initialize_on_open(view)
 
 
-def check_window_unloaded():
+def plugin_unloaded() -> None:
+    """Sublime Text API function called to unload this plugin."""
+    unload_settings()
+    unload_all_clients()
+
+
+def unload_all_clients() -> None:
     global clients_by_window
-    open_window_ids = list(window.id() for window in sublime.windows())
-    iterable_clients_by_window = clients_by_window.copy()
-    closed_windows = []
-    for id, window_clients in iterable_clients_by_window.items():
-        if id not in open_window_ids:
-            debug("window closed", id)
-            closed_windows.append(id)
+    for clients_by_config in clients_by_window.values():
+        for client in clients_by_config.values():
+            client.shutdown()
+    clients_by_window.clear()
+
+
+def check_window_unloaded() -> None:
+    open_window_ids = set(window.id() for window in sublime.windows())
+    closed_windows = set(id for id in clients_by_window if id not in open_window_ids)
     for closed_window_id in closed_windows:
         unload_window_clients(closed_window_id)
 
 
-def unload_window_clients(window_id: int):
+def unload_window_clients(window_id: int) -> None:
     global clients_by_window
-    if window_id in clients_by_window:
-        window_clients = clients_by_window[window_id]
-        del clients_by_window[window_id]
-        for config, client in window_clients.items():
-            debug("unloading client", config, client)
-            unload_client(client)
+    clients_by_config = clients_by_window.pop(window_id)
+    if clients_by_config:
+        for client in clients_by_config.values():
+            client.shutdown()
+        clients_by_config.clear()
 
 
-def unload_client(client: Client):
-    debug("unloading client", client)
-    try:
-        client.send_notification(Notification.exit())
-        client.kill()
-    except Exception as err:
-        exception_log("Error exiting server", err)
-
-
-def plugin_unloaded():
-    unload_settings()
-    for window in sublime.windows():
-        for client in window_clients(window).values():
-            unload_client(client)
+def unload_old_clients(window: sublime.Window):
+    project_path = get_project_path(window)
+    clients_by_config = window_clients(window)
+    clients_to_unload = []
+    for config_name, client in clients_by_config.items():
+        if client and client.get_project_path() != project_path:
+            debug('unload', config_name, 'project path changed from ', client.get_project_path())
+            clients_to_unload.append(config_name)
+    for config_name in clients_to_unload:
+        clients_by_config.pop(config_name).shutdown()
```

Comment:
Mypy should complain that pop() can return `None` here? (Not that this _will_ happen at runtime, but  it should reveal the issue anyways)

Reply:
Nope. Don't have any warning here. Neither in the diagnostics panel nor by running mypy directly.